src/t_list.c

```diff
@@ -394,76 +440,85 @@ void listElementsRemoved(client *c, robj *key, int where, robj *o) {
         dbDelete(c->db, key);
     }
     signalModifiedKey(c, c->db, key);
-    server.dirty++;
+    server.dirty += count;
 }
 
+/* Implements the generic list pop operation for LPOP/RPOP.
+ * The where argument specifies which end of the list is operated on. An
+ * optional count may be provided as the third argument of the client's
+ * command. */
 void popGenericCommand(client *c, int where) {
+    long count = 0;
+    robj *value;
+
+    if (c->argc > 3) {
+        addReplyErrorFormat(c,"wrong number of arguments for '%s' command",
+                            c->cmd->name);
+        return;
+    } else if (c->argc == 3) {
+        /* Parse the optional count argument. */
+        if (getPositiveLongFromObjectOrReply(c,c->argv[2],&count,NULL) != C_OK) 
+            return;
+        if (count == 0) {
+            /* Fast exit path. */
+            addReplyNullArray(c);
+            return;
+        }
+    }
+
     robj *o = lookupKeyWriteOrReply(c, c->argv[1], shared.null[c->resp]);
     if (o == NULL || checkType(c, o, OBJ_LIST))
         return;
 
-    robj *value = listTypePop(o, where);
-    if (value == NULL) {
-        addReplyNull(c);
-    } else {
+    if (!count) {
+        /* Pop a single element. This is POP's original behavior that replies
+        * with a bulk string. */
+        value = listTypePop(o,where);
+        serverAssert(value != NULL);
         addReplyBulk(c,value);
         decrRefCount(value);
-        listElementsRemoved(c,c->argv[1],where,o);
+        listElementsRemoved(c,c->argv[1],where,o,1);
+    } else {
+        /* Pop a range of elements. An addition to the original POP command,
+        *  which replies with a multi-bulk. */
+        int direction = (where == LIST_HEAD) ? LIST_TAIL : LIST_HEAD;
+        long llen = listTypeLength(o);
+        long rangelen = (count > llen) ? llen : count;
+        long rangestart = (direction == LIST_TAIL) ? 0 : -rangelen;
+        long rangeend = (direction == LIST_TAIL)
+                        ? rangelen - 1
+                        : -1;
+        if (o->encoding == OBJ_ENCODING_QUICKLIST) {
+            addListRangeReply(c,o,rangestart,rangeend,direction);
+            quicklistDelRange(o->ptr,rangestart,rangelen);
+            listElementsRemoved(c,c->argv[1],where,o,rangelen);
+        } else {
+            serverPanic("Unknown list encoding");
+        }
```

Comment:
IMHO the encoding check and panic can be dropped.
the call to addListRangeReply will already handle that.
logically, maybe the call for quicklistDelRange should be put inside that check, but i think it's pointless.
in our case, either add a single line assertion, before that block, or just drop it and rely on the check in addListRangeReply.
if we'll have more encodings in the future we can extend this code

Reply:
I kept it there because of quickListDelRange, but of course the addListRangeReply takes care of that. Perhaps another PR should take care of all these dangling panics.

Reply:
Dropped the check in bae0cb8